src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -888,70 +1019,221 @@ scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
 -- the sizes of the input runs. As as result, merge work would/could be more
 -- evenly distributed over time when the resident run is smaller than the worst
 -- case.
-scaleCreditsForMerge (MergingRun LevelLevelling _ _) c =
+scaleCreditsForMerge (MergingRun LevelLevelling _ _ _ _ _) (Credit c) =
     ScaledCredits (c * (1 + 4))
 -- A tiering merge has 5 runs at most (one could be held back to merged again)
 -- and must be completed before the level is full (once 4 more runs come in).
-scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _) c =
+scaleCreditsForMerge (MergingRun LevelTiering (NumRuns n) _ _ _ _) (Credit c) =
     ScaledCredits ((c * n + 3) `div` 4)
     -- same as division rounding up: ceiling (c * n / 4)
 
-{-# SPECIALISE supplyMergeCredits :: ScaledCredits -> MergingRun IO h -> IO () #-}
--- TODO: implement doing merge werk in batches, instead of always taking the
--- MVar. The thresholds for doing merge work should be different for each level,
--- maybe co-prime?
+{-# SPECIALISE supplyMergeCredits :: ScaledCredits -> CreditThreshold -> MergingRun IO h -> IO () #-}
+-- | Supply the given amount of credits to a merging run. This /may/ cause an
+-- ongoing merge to progress.
+--
+-- See Note [Credits]
 supplyMergeCredits ::
-     (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
+     forall m h. (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
   => ScaledCredits
+  -> CreditThreshold
   -> MergingRun m h
   -> m ()
-supplyMergeCredits _ SingleRun{} = pure ()
-supplyMergeCredits (ScaledCredits c) (MergingRun _ _ var) = do
-    mergeIsDone <- withMVar var $ \case
+supplyMergeCredits _ _ SingleRun{} = pure ()
+supplyMergeCredits
+  (ScaledCredits c)
+  thresh@(CreditThreshold creditsThresh)
+  (MergingRun _ _ numInputEntries unspentCreditsVar totalStepsVar mergeVar) = do
+    -- Add credits to the total of unspent credits
+    Credit unspentCredits <- addUnspentCredits unspentCreditsVar (Credit c)
+    -- Compute our /estimate/ of what the new total of unspent credits is. The
+    -- /actual/ total might have been changed again by a different thread, but
+    -- that is okay because we will find out if\/when we run takeCredits.
+    let unspentCredits' = unspentCredits + c
+
+    stepsDone <- readPrimVar (getTotalStepsVar totalStepsVar)
+
+    -- The merge is already finished
+    if stepsDone >= unNumEntries numInputEntries then
+      pure ()
```

Comment:
I'm trying to think about how we can *really* be confident that we can't for some reason have an unfinished merge, despite the number of steps performed matching the amount of work we had to do. :thinking: 

What happens, for example, if an async exception happens between `stepMerge` (returning True) and `completeMerge`? We did all the steps, but the merge will still remain ongoing and never be touched again.

Reply:
That's all true and the merge does get completed in the sense of doing all merge steps, but it can end up in a state where `Merge.complete` never gets called.

The scenario I am talking about it this:
- `stepMerge` succeeds and returns `True`, so `stepsDone` has been increased and the steps have actually been done, reaching the end of the merge (so `stepsDone` indeed does not exceed the actual amount of work done).
- an exception happens before `completeMerge` runs, so we remain in `OngoingMerge` state.
- `flushWriteBuffer` is not called (as you say)
- But in the *next update*, we enter `supplyMergeCredits` and find that `stepsDone` already must have been enough to complete the merge. So we simply call `pure ()`.
- When later `expectCompleteMerge` runs, it finds `OngoingMerge` (inside of which is a merge that has run to completion, but never been properly `complete`d). So it fails.

Reply:
Good point, that's indeed a case where it could find an `OngoingMerge`. So, maybe the solution is then to have a single mutable location holding a `False` bool that is set to `True` after `completeMerge`, and use that bool in this if condition